### PR TITLE
Default field values & default parameters

### PR DESCRIPTION
**Pull request by knikamo**
We have implemented two new features, _default field values_ and _default parameters_ during our Independent Project in Information Engineering (similar to a bachelor thesis).

**Default field values**
Default field values are assignments added at the top of the init-method during desugaring. Default field values does not work for the main class since the init method is not called. Tobias suggested that he would solve this. There is currently no logic to prevent circular references or other unwanted expressions.

**Default parameters**
For each default parameter in a function/method declaration a new function/method is added during desugaring. During typechecking all function/method calls are adjusted to call the function/method with correct number of arguments. Default parameters work as expected according to the tests we have added.

**General**
We have not done anything with error handling or error messages. For example the error message for wrong number of arguments is now a bit weird since a function/method not always expects a fixed number of arguments.

We have written a few tests for each feature, but probably they do not cover everything.

### CODE REVIEW

**Comment by supercooldave** [comment]
Awesome to see this. Very useful features to have in Encore.

File: src/types/Typechecker/Typechecker.hs
Comment (supercooldave) [inline src/types/Typechecker/Typechecker.hs]: What is Karro? 
If this is a comment, it could be removed.

**Comment by supercooldave** [comment]
There are a number of conflicts with the development branch of Encore. These need to be fixed before we can merge this PR. Do you need help doing that?

**Comment by kikofernandez** [comment]
hi there! You should do a rebase instead of a merge. If you are unsure about how to do it, let us know. (I am going to check how to fix your conflicts and see if I can push a rebased version. Does this sound good?)

**Comment by knikamo** [comment]
@kikofernandez aha okay it sounds good, I don't know how to do a rebase

File: src/parser/Parser/Parser.hs
Comment (kikofernandez) [inline src/parser/Parser/Parser.hs]: is this `returnWithEnd Field{fmeta, fmut, fname, ftypr, fexpr}` actually `returnWithEnd Field{fmeta, fmut, fname, ftype, fexpr}`  (misspelling of `ftypr` to `ftype`)

Comment (knikamo) [reply src/parser/Parser/Parser.hs]: yes!

**Comment by kikofernandez** [comment]
are you on the slack channel? There are some things that I need to explain and are probably not that relevant for the PR (otherwise send me an email to kiko.fernandez@it.uu.se and we continue some details that should be understood before I do this)

File: src/ir/AST/Desugarer.hs
Comment (kikofernandez) [inline src/ir/AST/Desugarer.hs]: not used. this can be removed

File: src/ir/AST/Desugarer.hs
Comment (kikofernandez) [inline src/ir/AST/Desugarer.hs]: should be removed, this is only for debugging

**Review by kikofernandez — CHANGES_REQUESTED** [review]
small number of improvements. Do not fix hem yet, as we are going to do the rebase and other stuff tomorrow.

**Comment by kikofernandez** [comment]
there is a binary file that can be removed too

**Review by kikofernandez — APPROVED** [review]

**Comment by knikamo** [comment]
@kikofernandez helped us with a rebase and fixing the conflicts. The test for default field values in the main class fails, @EliasC is responsible to fix that. 

Error messages are not adjusted to the new features. We are going to add issues about this after the PR is merged (right @EliasC ?)

**Comment by albertnetymk** [comment]
There was a small discussion on cleaning up the commit history a bit before merging. If there's still some pending changes you would like to do, I think it could be nice to add them while revising the history.

**Comment by EliasC** [comment]
> The test for default field values in the main class fails, @EliasC is responsible to fix that.

I have this implemented locally now and I believe that it should work out of the box. Should I push the commit directly to this branch?

> Error messages are not adjusted to the new features. We are going to add issues about this after the PR is merged (right @EliasC ?)

Yes please! The most important thing is that the known weaknesses are not forgotten. By creating an issue we increase the chances that someone is going to fix it at some point.

**Comment by kikofernandez** [comment]
@EliasC I think you should push directly to the branch. I do not want to re-write the history until everything that needs to be fixed has been fixed.

**Comment by kikofernandez** [comment]
Everyone: please let me know when everything is in place (issues have been created, all tests pass, etc) and I will do the interactive rebase to squash some commits together

**Comment by EliasC** [comment]
I've pushed my addition to make the final test case pass. If CI passes, I'm happy with my addition and will not add anything else. @knikamo gets the final word on when everything is in place.

**Comment by EliasC** [comment]
@knikamo: Can Kiko merge your commits and push it here? In other words, are you done and won't be adding anything else here?

**Comment by kikofernandez** [comment]
before merging, it would be good if I can get @casperstr and lowe's github email address, so that their commits are linked to their Github users. I'll wait until I get a confirmation on how to proceed

**Comment by kikofernandez** [comment]
I believe this is ready for merge @EliasC

**Comment by kikofernandez** [comment]
do not squash

**Comment by EliasC** [comment]
👍 @kikofernandez 

I am an author though, so you should press the button 😃

**Pull request merged**
